src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs

```diff
@@ -895,33 +898,66 @@ private static IEnumerable<PSTypeName> InferTypesFromCimCommand(PseudoBindingInf
             yield return new PSTypeName(typeof(CimInstance));
         }
 
-        private IEnumerable<PSTypeName> InferTypesFromForeachCommand(PseudoBindingInfo pseudoBinding)
+        private IEnumerable<PSTypeName> InferTypesFromForeachCommand(PseudoBindingInfo pseudoBinding, CommandAst commandAst)
         {
-            AstParameterArgumentPair argument;
-            if (pseudoBinding.BoundArguments.TryGetValue("Begin", out argument))
+            var previousPipelineElement = GetPreviousPipelineCommand(commandAst);
+            if (previousPipelineElement == null)
             {
-                foreach (var type in GetInferredTypeFromScriptBlockParameter(argument))
-                {
-                    yield return type;
-                }
+                yield break;
             }
-
-            if (pseudoBinding.BoundArguments.TryGetValue("Process", out argument))
+            foreach (var t in InferTypes(previousPipelineElement))
             {
-                foreach (var type in GetInferredTypeFromScriptBlockParameter(argument))
+                var thisToRestore = _context.CurrentThisType;
+                try
                 {
-                    yield return type;
-                }
-            }
+                    _context.CurrentThisType = t;
```

Comment:
I think this should be moved to where we recurse on the script method/property - you should know the type there as well, and it limits the scope more clearly.

You should also include some tests that depend on this value being set correctly.

Last - because it's effectively a global, you probably need a test that covers potential recursion where `$this` is used in two different script properties on completely different types and completion depends on correctly inferring the type both times.